File: Project.DbManagement/Entity/Voucher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using Project.DbManagement.Entity;
using System.ComponentModel.DataAnnotations.Schema;
using Project.DbManagement.Enum;

namespace Project.DbManagement
{
    public class Voucher: BaseEntity
    {
        [Key]
        public Guid Id { get; set; } 

        [Column(TypeName = "nvarchar(50)")]
        public string? Code { get; set; }

        [Column(TypeName = "nvarchar(256)")]
        public string VoucherName { get; set; }

        public string? ImageUrl { get; set; }

        public VocherTypeEnum? VoucherType { get; set; }
        public DateTime? StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public int? Status { get; set; }
        public decimal? DiscountAmount { get; set; }

        [Column(TypeName = "nvarchar(256)")]
        public string? Description { get; set; }

        public decimal? MinimumOrderAmount { get; set; }
        public decimal? DiscountPercentage { get; set; }

        //Tổng lượt sử dụng tối đa
        public int? TotalMaxUsage { get; set; }
        //Số lượt sử dụng tối đa trên mỗi khách hàng
        public int? MaxUsagePerCustomer { get; set; }
        //Số tiền giảm giá tối đa đối với voucher giảm theo phần trăm
        public decimal? MaxDiscountAmount { get; set; }
        //Tổng Số lượt đã sử dụng voucher
        public int? RedeemCount { get; set; }
    }
}

File: Fe/admin-site/src/types/voucher/voucher.ts
import { PaginationParams } from "../common/pagination";

export default interface VoucherReqDto {
  id: string;
  voucherName: string;
  voucherType: number;
  startDate: string;
  endDate: string;
  createdByUserId: string;
  lastModifiedByUserId: string;
  lastModifiedOnDate: string;
  createdOnDate: string;
  status: number;
  isDeleted: boolean;
  code: string; // Mã voucher
  discountAmount: number | null; // Số tiền giảm giá
  discountPercentage: number | null; // Phần trăm giảm giá
  description: string | null; // Mô tả voucher
  minimumOrderAmount: number | null; // Giá trị đơn hàng tối thiểu
  totalMaxUsage: number | null; //Tổng lượt sử dụng tối đa
  maxUsagePerCustomer: number | null; //Số lượt sử dụng tối đa trên mỗi khách hàng
  maxDiscountAmount: number | null; //Số tiền giảm giá tối đa đối với voucher giảm theo phần trăm
  redeemCount: number | null; //Tổng Số lượt đã sử dụng voucher
  imageUrl: string | null; // Đường dẫn đến hình ảnh của voucher
}

export interface VoucherResDto {
  id: string;
  voucherName: string;
  status: number | null;
  voucherType: number | null;
  createdOnDate: string;
  lastModifiedOnDate: string;
  createdByUserId: string;
  lastModifiedByUserId: string;
  startDate: string;
  endDate: string;
  isDeleted: boolean;
  code: string; // Mã voucher
  discountAmount: number | null; // Số tiền giảm giá
  discountPercentage: number | null; // Phần trăm giảm giá
  description: string | null; // Mô tả voucher
  minimumOrderAmount: number | null; // Giá trị đơn hàng tối thiểu
  totalMaxUsage: number | null; //Tổng lượt sử dụng tối đa
  maxUsagePerCustomer: number | null; //Số lượt sử dụng tối đa trên mỗi khách hàng
  maxDiscountAmount: number | null; //Số tiền giảm giá tối đa đối với voucher giảm theo phần trăm
  redeemCount: number | null; //Tổng Số lượt đã sử dụng voucher
  imageUrl: string | null; // Đường dẫn đến hình ảnh của voucher
}

export interface VoucherFilterParams extends PaginationParams {
  ten_giam_gia?: string;
  statusTotal?: number; //Trạng thái tổng: Dựa vào EndDate StartDate và Status để quyết định
}
